dlib/dnn/cuda_dlib.cu

```diff
@@ -1172,6 +1177,476 @@ namespace dlib
 
     // ----------------------------------------------------------------------------------------
 
+        __global__ void _cuda_nn_upsample(float* out, const float* in, size_t nr, 
+                                          size_t nc, 
+                                          size_t nc2,
+                                          size_t scale_y, 
+                                          size_t scale_x)
+        {
+
+            for (auto r : grid_stride_range_y(0, nr))
+            {
+                for (auto c : grid_stride_range(0, nc))
+                {
+                    auto i = r * nc + c;                    
+                    auto j = (r / scale_y) * nc2 + c/scale_x;                    
+                    out[i] = in[j];
+                }
+            }
+        }
+
+        __global__ void _cuda_bilinear_upsample(float* out, const float* in, size_t nr, 
+                                                size_t nc, 
+                                                size_t nr2,                                                
+                                                size_t nc2,
+                                                size_t scale_y, 
+                                                size_t scale_x)
+        {
+
+            for (auto r : grid_stride_range_y(0, nr))
+            {
+                auto srcR = r/scale_y;
+                auto srcP = in + srcR * nc2;
```

Comment:
I just looked at the CUDA code.  These kernels would be a lot faster if, rather than launching many of them you processed the entire tensor with one grid stride loop.  You also don't need the grid_stride_range_y loop.  You will get better hardware occupancy with one grid stride loop over everything.  The only time you need 2D grid stride loops is when you need to do something complex that has hardware limits on how the warps are structured.  For instance, I think the only time I've needed 2D grid stride loops is when doing complex warp reductions, which you aren't using here (and don't need).